### PR TITLE
Refactor volume.Builder.IsReadOnly to GetReadWriteMode

### LINKED ISSUE #16514: Add a way to distinguish between volumes that are synthesized by the kubelet and 'real' volumes
Currently in the ownership and SELinux label management code, we do not alter volumes whose builder reports `true` from `IsReadOnly()`.  This doesn't give us quite the right decision plane; we need a new method that tells the kubelet where a volume is synthesized by the system or not.  I suggest:

```go
package volume

type Builder interface {
  // other methods omitted
  IsSynthetic() bool
}
```

`secrets` and `downwardApi` would return 'true' from this method.

Then the decision plane for ownership / selinux label management becomes:

```go
if builder.IsReadOnly() && !builder.IsSynthetic() {
  // do stuff
}
```

thoughs @thockin @swagiaal @pweil- @smarterclayton ?

### PR DESCRIPTION
**Pull request by pmorie**
Resolves #16514.

### CODE REVIEW

**Comment by pmorie** [comment]
@vishh @thockin @swagiaal @pweil- PTAL

File: pkg/volume/volume.go [hunk 1]
```diff
@@ -103,3 +113,31 @@ func RenameDirectory(oldPath, newName string) (string, error) {
 	}
 	return newPath, nil
 }
+
+// ReadWriteModeFor returns the ReadWriteMode represented by the given boolean.
+// A value of true maps to ReadOnly, false to ReadWrite
+func ReadWriteModeFor(ro bool) ReadWriteMode {
```
Comment (thockin) [inline pkg/volume/volume.go]: This is a sketchy function.  Is it really needed?

File: pkg/volume/volume.go [hunk 2]
```diff
@@ -103,3 +113,31 @@ func RenameDirectory(oldPath, newName string) (string, error) {
...
+	if ro {
+		return ReadOnly
+	}
+
+	return ReadWrite
+}
+
+func IsReadOnlyToPod(mode ReadWriteMode) bool {
```
Comment (thockin) [inline pkg/volume/volume.go]: I would pass a Builder here and call GetReadWriteMode() internally

File: pkg/volume/aws_ebs/aws_ebs.go [hunk 3]
```diff
@@ -177,10 +177,18 @@ type awsElasticBlockStoreBuilder struct {
 
 var _ volume.Builder = &awsElasticBlockStoreBuilder{}
 
+func (b *awsElasticBlockStoreBuilder) GetReadWriteMode() volume.ReadWriteMode {
+	return volume.ReadWriteModeFor(b.readOnly)
```
Comment (thockin) [inline pkg/volume/aws_ebs/aws_ebs.go]: maybe ReadWriteModeFromBool() so it's more obvious

**Comment by thockin** [comment]
Just a few nits.  I am travelling tomorrow, so others can final approve

File: pkg/volume/volume.go (hunk 1 above)
Comment (pmorie) [inline pkg/volume/volume.go]: It is needed to bridge the ReadOnly boolean on volumes.  I don't think we want to expose `ReadWriteMode` to users.

File: pkg/volume/volume.go [hunk 4]
```diff
@@ -30,6 +30,15 @@ type Volume interface {
 	GetPath() string
 }
 
+// ReadWriteMode represents the usage mode for a volume.
+type ReadWriteMode string
```
Comment (vishh) [inline pkg/volume/volume.go]: Do we want to name this `FsPermissions` instead? This type is conveying more than just `read` or `read-write` modes. For example `managed` mode.

How about a struct with `permissions` and `management` mode?

File: pkg/volume/volume.go [hunk 5]
```diff
@@ -42,9 +51,10 @@ type Builder interface {
 	// directory path, which may or may not exist yet.  This may be called
 	// more than once, so implementations must be idempotent.
 	SetUpAt(dir string) error
-	// IsReadOnly is a flag that gives the builder's ReadOnly attribute.
-	// All persistent volumes have a private readOnly flag in their builders.
-	IsReadOnly() bool
+	// GetReadWriteMode returns the ReadWriteMode of the volume.
+	// If the mode is ReadWrite or ReadOnlyManaged, the Kubelet will perform
+	// ownership and selinux label management on appropriate volumes.
```
Comment (vishh) [inline pkg/volume/volume.go]: nit: I wouldn't expect this interface to contain information about how the user of this interface `kubelet` will be using this information.

File: pkg/volume/volume.go (hunk 1 above)
Comment (vishh) [inline pkg/volume/volume.go]: I find this sketchy as well. Is it possible to abstract this into a base class that volume builders will instantiate and use?

File: pkg/volume/volume.go (hunk 2 above)
Comment (vishh) [inline pkg/volume/volume.go]: Question: Does the volumes plugin really have to understand pods and nodes? Can this policy be scoped at the kubelet level? Ideally, we can use these plugins even outside of the kubelet context right?

**Comment by vishh** [comment]
I posted some minor code health comments. The basic structure LGTM.

**Comment by pmorie** [comment]
@thockin @vishh Thanks for the feedback -- I'm working on it today.  I found one little but I'm going to make this fix as well.

**Comment by pmorie** [comment]
@vishh I missed your earlier points about factoring of the type, my bad.  I would be okay for renaming to `FsPermissions`, I'm not sure it's absolutely warranted yet for a struct with more fields.  Are you okay with leaving the factoring as-is for now, and we can refactor later if needed?

**Comment by pmorie** [comment]
@thockin @vishh I think I missed most of vish's comments somehow.  I'll take another pass through this PR and ping you when it's ready for review.

**Comment by vishh** [comment]
@pmorie: The existing `string` type is mildly confusing already. I'm concerned that we will overload this type with those attributes in the future. I'd be less concerned if we expose an interface and use whatever representation underneath.

**Comment by pmorie** [comment]
@vishh 

> The existing string type is mildly confusing already. I'm concerned that we will overload this type with those attributes in the future. I'd be less concerned if we expose an interface and use whatever representation underneath

Accepted.  How about:

```go
type Attributes struct {
  ReadOnly bool
  Managed  bool
}

type Builder interface {
  GetAttributes() Attributes
}
```

Unless we want to attach behaviors to this

**Comment by vishh** [comment]
What's your opinion on using interfaces instead for attributes? Will it be
helpful at all?

```go
type Attributes interface {
   ReadOnly() bool
   Managed() bool
   ... // What ever else in the future.
}
```

**Comment by pmorie** [comment]
@vishh Personally, my subjective taste leans towards using a struct unless we want to attach behaviors or differentiate implementations.  The primary reason I prefer the struct is because I don't think it will be helpful to attach behaviors or differentiate implementations.

Thoughts?  I'll proceed w/ a struct as follows later this weekend. I'm flying on Sunday for Kubecon and probably won't have time until I'm on a plane to do this:

```go
type Attributes struct {
  ReadOnly bool
  Managed  bool
  SupportsOwnershipManagement    bool
  SupportsSELinuxLabelManagement bool
}
```

I'll note that refactoring to the above makes it possible to easily implement opt-out features for using `FSGroup` with certain volumes if the need arises.

**Comment by vishh** [comment]
SGTM. Thanks!

**Comment by pmorie** [comment]
@vishh Refactored as discussed -- PTAL.

**Comment by vishh** [comment]
LGTM. Thanks @pmorie!

**Comment by pmorie** [comment]
@vishh Unfortunately I had forgotten to add `[Skipped]` to the e2e I added for this, so you will need to retag :(

**Comment by vishh** [comment]
@pmorie: unit tests are still failing. Ping me once you have fixed them.

**Comment by pmorie** [comment]
@vishh hrm, my bad -- weirdly, this is what I see next to unit/integration:

```
Jenkins unit/integration — 2231 tests run, 8 skipped, 0 failed.
```

**Comment by vishh** [comment]
@pmorie: pkg/kubelet build is failing.

```sh
16:38:45 _output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_test.go:516: undefined: volume.ReadWriteMode
16:38:45 _output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_test.go:517: undefined: volume.ReadWrite
16:38:45 _output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_test.go:563: cannot use stubVolume literal (type *stubVolume) as type volume.Builder in field value:
16:38:45 	*stubVolume does not implement volume.Builder (missing GetAttributes method)
16:38:45 _output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_test.go:564: cannot use stubVolume literal (type *stubVolume) as type volume.Builder in field value:
16:38:45 	*stubVolume does not implement volume.Builder (missing GetAttributes method)
16:38:45 _output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_test.go:565: cannot use stubVolume literal (type *stubVolume) as type volume.Builder in field value:
16:38:45 	*stubVolume does not implement volume.Builder (missing GetAttributes method)
```

**Comment by pmorie** [comment]
@vishh weird, I will fix.

**Comment by pmorie** [comment]
@vishh PTAL

**Comment by pmorie** [comment]
@vishh Should be good now

**Comment by pmorie** [comment]
Thanks for the reviews, @vishh

**Pull request merged**